### ISSUE TITLE
Pydantic v2

### ISSUE THREAD

**Issue by courtland** [sabuhish/fastapi-mqtt#68]
Pydantic 2.0.x was recently released, which is [now supported by FastAPI](https://github.com/tiangolo/fastapi/releases/tag/0.100.0). 

It would be great to release a version of this package that permits pydantic > 2.0. 

There is a script in the [migration guide](https://docs.pydantic.dev/2.0/migration/) that tries to automatically convert code.
I think the only issue for fastapi-mqtt would be that [BaseSettings was moved to a separate package](https://docs.pydantic.dev/2.0/migration/#basesettings-has-moved-to-pydantic-settings).

**Comment by tim-timman** [comment]
Due to the requirements specification not allowing is to specify that we need one or the other of `pydantic >=1.9.0,<2.0.0`, `pydantic-settings >=2.0.1` – unless we want to take a stance to only support the newer – we need some "graceful" workaround.

My suggestion is to not have either as a dependency and have a fallback for what I _believe_ is the common case – to just use it as shown in the docs an examples. The features BaseSettings provides isn't really utilized (by fastapi-mqtt at least), and its only in [`config.py`](https://github.com/sabuhish/fastapi-mqtt/blob/master/fastapi_mqtt/config.py).

Here some code of what I think it could look like compared to the [current suggestion][pr-diff]:

[pr-diff]: https://github.com/sabuhish/fastapi-mqtt/pull/67/files#diff-09f6ef8188c8b1114bfac0b5cdf09d509837f1129d796cc6fea78c74ad2f9f5fR6-R11

```python
try:
    from pydantic_settings import Settings
except ImportError:
    try:
        from pydantic import Settings
    except ImportError:
       import warnings
        message = (
            "Neither package ('pydantic-settings >=2.0.1', 'pydantic >=1.9.0,<2.0.0') is installed that provide BaseSettings.\n"
            "Please install one of those packages for intended function (config from file, environment, etc).\n"
            "NOTE! Falling back to a simple, feature incomplete implementation. Ignore at your own risk!"
        )
        warnings.warn(message)
        exception = ValueError(f"Something's probably not correct. This is only fallback validation.\n\n{message}")
        class Settings:
            def __init__(self, **kwargs):
                aggregated_annotations = {}
                for cls in type(self).__mro__[1::-1]:
                    aggregated_annotations.update(getattr(cls, "__annotations__", {}))
                self.__dict__["__annotations__"] = aggregated_annotations
                hints = self.__annotations__.copy()
                missing = object()
                for key, value in kwargs.items():
                    if hints.pop(k, missing) is missing:
                        raise exception
                    self.__dict__[key] = value
                for key in hints:
                    if getattr(self, key, missing) is missing:
                        raise exception

            def __setattr__(self, key, value):
                if key not in self.__annotations__:
                    raise exception
                self.__dict__[key] = value
```

One could simply issue the warning and use `types.SimpleNamespace`, but then there's no validation whatsoever. A typo by a developer wouldn't raise an error, and be hard to track down (especially if the warning is missed).

Raise an error if neither are present or drop the dependency on BaseSettings all together are also options.

Any thoughts @mblo @azogue @sabuhish ?

**Comment by tim-timman** [comment]
@azogue I like your suggestion of just moving to using BaseModel from pydantic instead of BaseSettings. I think that's the right move.. 

Those features aren't in use for this package and I doubt it's used by any user. Though, if someone is and _need_ them, wouldn't extending it yourself work? (not able to test in some time)

```python
from fastapi_mqtt import MQTTConfig as _MQTTConfig
from pydantic_settings import BaseSettings  # or from pydantic

class MQTTConfig(BaseSettings, _MQTTConfig):
    pass
```
Then just use it as it's is today. I think the MRO will end up the same since BaseSettings itself extends BaseModel. So unless the metaclass does some really wild stuff, I think it would work the same. If it works as intended, adding this snippet of how to opt-in to the old way in the docs and switch to just BaseModel ourselves is the best way forward.

**Comment by azogue** [comment]
> Don't know what @sabuhish want to do here 😜, but I offer myself to make those changes asap (either set the pyproject for those v1.2 and v2.0 versions, or explore the migration from BaseSettings to BaseModel) 👍

* In #69, drop BaseSettings usage and support pydantic v1 & v2, with version 1.2.0
* In #70, require pydantic v2 and update deps in lock file, with version 2.0

**Comment by tim-timman** [comment]
> ```python
> from fastapi_mqtt import MQTTConfig as _MQTTConfig
> from pydantic_settings import BaseSettings  # or from pydantic
> 
> class MQTTConfig(BaseSettings, _MQTTConfig):
>     pass
> ```

Just tested this, and it seems to work as expected. So maybe add it to the docs in the PR's in case someone updating want an easy way to utilize the old behavior? @azogue 

Maybe add to README.md something like the following (**Note:** remove `/` from lines with tripple backticks):

```markdown
### Still need pydantic's BaseSettings for MQTTConfig after >=1.2.0?

\```python
from fastapi_mqtt import MQTTConfig as _MQTTConfig
# Uncomment the import for the specific version of pydantic you're using
# from pydantic_settings import BaseSettings  # Using pydantic v2, also install pydantic-settings
# from pydantic import BaseSettings           # Using pydantic v1

class MQTTConfig(BaseSettings, _MQTTConfig):
    """extends MQTTConfig to use pydantic's BaseSettings features"""
    pass
\```
Snippet for making the MQTTConfig class utilize pydantic's BaseSettings functionality.

The MQTTConfig class was changed in version 1.2.0 to use pydantic's BaseModel instead of BaseSettings to make it compatible with both v1 and v2. In v2 BaseSettings was moved to its own package [`pydantic_settings`](https://pypi.org/project/pydantic-settings/) and this made it impossible to specify requirements for this package without choosing to only support one or the other. `fastapi-mqtt` itself didn't utilize any BaseSettings specific functionality, nor was it mentioned in its documentation and examples – do the probability of anyone depending on them was deemed small. If you did use it, or want to, use small snippet provided above.
```

> * In [Use pydantic BaseModel for MQTTConfig object #69](https://github.com/sabuhish/fastapi-mqtt/pull/69), drop BaseSettings usage and support pydantic v1 & v2, with version 1.2.0

I am reconsidering if doing this breaking change (even though low probability) in a minor version is the right thing to do, since there is the possibility it is depended upon. Is it more beneficial to create both the 1.2.0 and 2.0.0 versions somehow that I'm not seeing?

**Comment by azogue** [comment]
@tim-timman, I don't see the need of that 'migration' instructions, as we agreed that usage of `BaseSettings` in this library was unnecessary \*\*

With some short notice in the release of v1.2 I think it would be enough to document the (non-breaking) change. But, anyway, it's for @sabuhish to decide what to do 🤷

---

\*\* In fact, being this a _plugin_ for `fastapi`, I suppose that most of the usages of Pydantic's `BaseSettings` to store the app config will contain some field with the MQTT settings, and in that case, it's the same to use a `BaseModel` or a `BaseSettings` dataclass to contain the nested data for the plugin config 👍

```python
class Settings(BaseSettings):
    """FastAPI app settings."""
    # example app config fields:
    app_port: int = 8081
    app_log_json: bool = False
    app_log_level: str = "INFO"
    # if mqtt is optional:
    mqtt_opc_config: MQTTSettings | None = None
    # if mqtt is required
    mqtt_config: MQTTSettings = Field(default_factory=MQTTSettings)
```

**Comment by tim-timman** [comment]
@azogue fair enough.

@sabuhish any way we can further facilitate getting something like this merged and published?

**Comment by sabuhish** [comment]
This feature has already been released in version 1.2.0. Thanks, everyone for their efforts here!
